src/MSBuildLocatorAnalyzer/MSBuildLocatorAnalyzerAnalyzer.cs

```diff
@@ -0,0 +1,75 @@
+        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(Resources.AnalyzerMessageFormat), Resources.ResourceManager, typeof(Resources));
+        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(Resources.AnalyzerDescription), Resources.ResourceManager, typeof(Resources));
+        private const string Category = "Naming";
+
+        private readonly string[] msBuildAssemblies =
+        {
+            "Microsoft.Build",
+            "Microsoft.Build.Engine",
+            "Microsoft.Build.Framework",
+            "Microsoft.Build.Tasks.Core",
+            "Microsoft.Build.Utilities.Core"
+        };
+
+        private readonly string[] msbuildLocatorRegisterMethods =
+        {
+            "RegisterDefaults",
+            "RegisterInstance",
+            "RegisterMSBuildPath"
+        };
+
+        private static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Error, isEnabledByDefault: true, description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterOperationAction(opact =>
+            {
+                var invocation = (IInvocationOperation)opact.Operation;
+                var targetMethod = invocation.TargetMethod;
+                if (targetMethod?.ContainingAssembly != null)
+                {
+                    if (targetMethod.ContainingAssembly.Name.Equals("Microsoft.Build.Locator") && msbuildLocatorRegisterMethods.Contains(targetMethod.Name))
+                    {
+                        MethodDeclarationSyntax method = (MethodDeclarationSyntax)invocation.Syntax.FirstAncestorOrSelf((Func<SyntaxNode, bool>)(a => a is MethodDeclarationSyntax));
+                        if (method != null)
+                        {
+                            IOperation symbolParent;
+                            for (symbolParent = invocation; symbolParent.Parent != null && symbolParent.Kind != OperationKind.MethodBody; symbolParent = symbolParent.Parent);
+                            foreach (var child in symbolParent.Descendants())
+                            {
+                                if ((child is IInvocationOperation op && msBuildAssemblies.Contains(op.TargetMethod?.ContainingAssembly?.Name)) ||
+                                (child is ITypeOfOperation toOp && msBuildAssemblies.Contains(toOp.TypeOperand.ContainingAssembly?.Name)) ||
+                                (msBuildAssemblies.Contains(child.Type?.ContainingAssembly?.Name)))
+                                {
+                                    opact.ReportDiagnostic(Diagnostic.Create(Rule, child.Syntax.GetLocation()));
+                                }
+                            }
```

Comment:
A different approach would be something like this:

```csharp
context.RegisterOperationBlockStartAction(context =>
{
    if (context.OwningSymbol.Kind != SymbolKind.Method)
    {
        return;
    }

    var hasRegisterMethod = false;
    var hasBadReference = false;
    
    context.RegisterOperationAction(context =>
    {
        var invocation = (IInvocationOperation)opact.Operation;
        var targetMethod = invocation.TargetMethod;
        if (targetMethod.ContainingAssembly.Name.Equals("Microsoft.Build.Locator") && msbuildLocatorRegisterMethods.Contains(targetMethod.Name))
        {
            hasRegisterMethod = true;
        }
        else if (msBuildAssemblies.Contains(targetMethod.ContainingAssembly.Name)))
        {
            hasBadReference = true;
        }
    }, OperationKind.Invocation);

    context.RegisterOperationAction(context =>
    {
        var type = context.Operation switch
        {
            IObjectCreationOperation objectCreation => objectCreation.Type,
            ITypeOfOperation typeofOperation => typeofOperation.TypeOperand,
            _ => null,
        };

        if (msBuildAssemblies.Contains(type?.ContainingAssembly.Name)))
        {
            hasBadReference = true;
        }
    }, OperationKind.ObjectCreation, OperationKind.TypeOf);

    context.RegisterOperationBlockEndAction(context =>
    {
        if (hasRegisterMethod && hasBadReference)
        {
            // report diagnostic
        }
    });
});
```

Not sure if that's better. I wrote directly in GitHub so could have made mistakes, but the idea is to first register operation block start action, then on that, register operation actions for invocation, typeof, and object creation and gather the needed info (whether a register method is called and whether there is another problematic reference)

Then, on operation block end, we have gathered the needed info, and decide whether to report a diagnostic or not.